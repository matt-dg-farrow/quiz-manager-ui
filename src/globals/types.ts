export type Answer = { id: number; text: string; isCorrect: boolean };
export type Question = { id: number; title: string; answers: Answer[] };
export type Quiz = { id: number; title: string; questions: Question[] };
