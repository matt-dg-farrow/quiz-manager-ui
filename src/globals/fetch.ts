import { Quiz } from "./types";

const API_BASE = "http://localhost:8080/";

const headers = {
  "Content-Type": "application/json; charset=utf-8",
};

export const Post = (url: any, query: any) =>
  fetch(`${API_BASE}${url}`, {
    headers,
    method: "POST",
    credentials: "same-origin",
    body: JSON.stringify(query),
  })
    .then((resp) => resp.json())
    .catch(() => ({
      ok: false,
      data: { statusMessage: "Parsing POST Error" },
    }));

export const Get = (url: any, query?: any) =>
  fetch(`${API_BASE}${url}`, {
    headers,
    method: "GET",
    credentials: "same-origin",
    body: JSON.stringify(query),
  })
    .then((resp) => resp.json())
    .catch(() => ({ ok: false, data: { statusMessage: "Parsing GET Error" } }));

export const Delete = (url: any, query: any) =>
  fetch(`${API_BASE}${url}`, {
    headers,
    method: "DELETE",
    credentials: "same-origin",
    body: JSON.stringify(query),
  })
    .then((resp) => resp.json())
    .catch(() => ({
      ok: false,
      data: { statusMessage: "Parsing DELTE Error" },
    }));

export const Put = (url: any, query: any) =>
  fetch(`${API_BASE}${url}`, {
    headers,
    method: "PUT",
    credentials: "same-origin",
    body: JSON.stringify(query),
  })
    .then((resp) => resp.json())
    .catch(() => ({
      ok: false,
      data: { statusMessage: "Parsing PUT Error" },
    }));
