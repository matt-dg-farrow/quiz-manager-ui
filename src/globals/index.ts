export enum QuizRequestTypes {
  FETCH_QUIZ_DATA_ACTION = "FETCH_QUIZ_DATA_ACTION",
  FETCH_QUIZ_DATA_ACTION_SUCCESS = "FETCH_QUIZ_DATA_ACTION_SUCCESS",
  FETCH_QUIZ_DATA_ACTION_FAILURE = "FETCH_QUIZ_DATA_ACTION_FAILURE",
  FETCH_QUIZ_DATA_REQUEST = "quiz/all",

  CREATE_QUIZ_ACTION = "CREATE_QUIZ_ACTION",
  CREATE_QUIZ_ACTION_SUCCESS = "CREATE_QUIZ_ACTION_SUCCESS",
  CREATE_QUIZ_ACTION_FAILURE = "CREATE_QUIZ_ACTION_FAILURE",
  CREATE_QUIZ_ACTION_REQUEST = "quiz/create",

  DELETE_QUIZ_ACTION = "DELETE_QUIZ_ACTION",
  DELETE_QUIZ_ACTION_SUCCESS = "DELETE_QUIZ_ACTION_SUCCESS",
  DELETE_QUIZ_ACTION_FAILURE = "DELETE_QUIZ_ACTION_FAILURE",
  DELETE_QUIZ_ACTION_REQUEST = "quiz/delete",

  EDIT_QUIZ_ACTION = "EDIT_QUIZ_ACTION",
  EDIT_QUIZ_ACTION_SUCCESS = "EDIT_QUIZ_ACTION_SUCCESS",
  EDIT_QUIZ_ACTION_FAILURE = "EDIT_QUIZ_ACTION_FAILURE",
  EDIT_QUIZ_ACTION_REQUEST = "quiz/update",
}
