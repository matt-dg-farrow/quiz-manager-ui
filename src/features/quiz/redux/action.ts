import { createAction } from "@reduxjs/toolkit";
import { QuizRequestTypes } from "../../../globals";
import { Quiz } from "../../../globals/types";

export const FetchQuizDataAction = createAction(
  QuizRequestTypes.FETCH_QUIZ_DATA_ACTION
);

export const FetchQuizDataActionSuccess = createAction<Quiz[]>(
  QuizRequestTypes.FETCH_QUIZ_DATA_ACTION_SUCCESS
);

export const FetchQuizDataActionFailure = createAction<any>(
  QuizRequestTypes.FETCH_QUIZ_DATA_ACTION_FAILURE
);

export const CreateQuizAction = createAction<Quiz>(
  QuizRequestTypes.CREATE_QUIZ_ACTION
);

export const CreateQuizActionSuccess = createAction(
  //does this need a payload?
  QuizRequestTypes.CREATE_QUIZ_ACTION_SUCCESS
);

export const CreateQuizActionFailure = createAction<any>(
  //does this need a payload?
  QuizRequestTypes.CREATE_QUIZ_ACTION_FAILURE
);

export const DeleteQuizAction = createAction<Quiz>(
  QuizRequestTypes.DELETE_QUIZ_ACTION
);

export const DeleteQuizActionSuccess = createAction(
  //does this need a payload?
  QuizRequestTypes.DELETE_QUIZ_ACTION_SUCCESS
);

export const DeleteQuizActionFailure = createAction<any>(
  //does this need a payload?
  QuizRequestTypes.DELETE_QUIZ_ACTION_FAILURE
);

export const EditQuizAction = createAction<Quiz>(
  QuizRequestTypes.EDIT_QUIZ_ACTION
);

export const EditQuizActionSuccess = createAction(
  //does this need a payload?
  QuizRequestTypes.EDIT_QUIZ_ACTION_SUCCESS
);

export const EditQuizActionFailure = createAction<any>(
  //does this need a payload?
  QuizRequestTypes.EDIT_QUIZ_ACTION_FAILURE
);
