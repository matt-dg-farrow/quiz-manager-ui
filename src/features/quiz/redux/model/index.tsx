import { Quiz } from "../../../../globals/types";
import {
  FetchQuizDataAction,
  FetchQuizDataActionFailure,
  FetchQuizDataActionSuccess,
} from "../action";

export interface IFetchQuizInitialState {
  quizzes: Quiz[];
}

export interface ICreateQuizPayload {
  quiz: Quiz;
}

export interface ICreateQuizInitialState {
  payload: ICreateQuizPayload;
}

export interface IDeleteQuizPayload {
  quiz: Quiz;
}

export interface IDeleteQuizInitialState {
  payload: IDeleteQuizPayload;
}

export type FetchQuizRequestAction =
  | typeof FetchQuizDataAction
  | typeof FetchQuizDataActionSuccess
  | typeof FetchQuizDataActionFailure;
