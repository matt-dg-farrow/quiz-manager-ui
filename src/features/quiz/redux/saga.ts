import { call, put, takeLatest } from "redux-saga/effects";
import { QuizRequestTypes } from "../../../globals";
import { Get } from "../../../globals/fetch";
import {
  FetchQuizDataActionFailure,
  FetchQuizDataActionSuccess,
} from "./action";

export function* fetchQuizDataWorkerSaga() {
  try {
    // @ts-ignore
    const response = yield call(Get, QuizRequestTypes.FETCH_QUIZ_DATA_REQUEST);
    yield put(FetchQuizDataActionSuccess(response));
  } catch (error) {
    yield put(FetchQuizDataActionFailure(error));
  }
}

export default function* watcherSaga(): any {
  yield takeLatest(
    QuizRequestTypes.FETCH_QUIZ_DATA_ACTION,
    fetchQuizDataWorkerSaga
  );
}
