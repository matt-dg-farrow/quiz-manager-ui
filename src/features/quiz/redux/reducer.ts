import { createSlice } from "@reduxjs/toolkit";
import { Quiz } from "../../../globals/types";
import {
  CreateQuizAction,
  FetchQuizDataAction,
  FetchQuizDataActionSuccess,
} from "./action";
import { ICreateQuizPayload, IFetchQuizInitialState } from "./model";

const fetchInitialState: IFetchQuizInitialState = {
  quizzes: [],
};

const fetchQuizDataSlice = createSlice({
  name: "quiz",
  initialState: fetchInitialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(FetchQuizDataActionSuccess, (state, action) => {
      state.quizzes = action.payload;
    });
  },
});

export const FetchQuizDataReducer = fetchQuizDataSlice.reducer;
