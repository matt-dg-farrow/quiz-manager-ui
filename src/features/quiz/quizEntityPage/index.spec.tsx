import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { Provider } from "react-redux";
import QuizEntityPage from ".";
import configureStore from "redux-mock-store";
import { Quiz } from "../../../globals/types";
import { IFetchQuizInitialState } from "../redux/model";

describe("The QuizEntityPage component", () => {
  const actionPayload: Quiz = {
    id: 1,
    title: "Quiz 1",
    questions: [
      {
        id: 2,
        title: "Question 1",
        answers: [
          { id: 3, text: "Answer 1", isCorrect: true },
          { id: 4, text: "Answer 1", isCorrect: true },
          { id: 5, text: "Answer 1", isCorrect: true },
        ],
      },
    ],
  };

  const populatedInitialState: IFetchQuizInitialState = {
    quizzes: [actionPayload],
  };

  const initialStoreState = { FetchQuizDataReducer: populatedInitialState };

  const mockStore = configureStore();
  const store = mockStore(initialStoreState);
  const wrapper = shallow(
    <Provider store={store}>
      <QuizEntityPage quizId="1" />
    </Provider>
  );

  it("should contain the correct number of components", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
