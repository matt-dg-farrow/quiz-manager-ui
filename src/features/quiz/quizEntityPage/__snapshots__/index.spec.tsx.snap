// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The QuizEntityPage component should contain the correct number of components 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <QuizEntityPage
    quizId="1"
  />
</ContextProvider>
`;
